### ISSUE TITLE
Detecting sky in movement code?

### ISSUE THREAD

**Issue by johnstt** [ValveSoftware/halflife#1866]
I've got some code that uses `pmove->PM_PlayerTrace` to look at the area behind the player (to see what they can walljump from).

I want to prevent walljumping off the sky, so I tried calling `pmove->PM_PointContents(trace.endpos,NULL)`.

This seems to give me -1 (`CONTENTS_EMPTY`) everywhere except water (-3 : `CONTENTS_WATER`). So...does it just not work or what? Is there a better way?

I have tons of code in the server end of things that uses `UTIL_PointContents` and it works, but I'm trying to do this using what's already available in `pm_shared.c`...

**Comment by SamVanheer** [comment]
That function is a bit weird, the return value isn't the contents value it's some constrained value that is almost always `CONTENTS_EMPTY`, `CONTENTS_SOLID` or `CONTENTS_WATER`. The real contents value is the second parameter. Pass it in to get that:
```
int contents;
pmove->PM_PointContents( trace.endpos, &contents );
```

This value is always set so you can assume it's valid after the call.

**Comment by SamVanheer** [comment]
You're sure you're checking the variable you passed as a parameter and not the return value, right? The code shows it's getting the value that was read from the BSP file, so unless the engine's code is doing something wrong it should work.

**Comment by SamVanheer** [comment]
Maybe the maps were compiled with a compiler that changed CONTENTS_SKY to CONTENTS_SOLID for some reason?

**Comment by johnstt** [comment]
Is there a way to check the texture instead? (it's always `SKY` right?)
arggh.

**Comment by SamVanheer** [comment]
You could try using PM_TruePointContents, it should return exactly what the truecontents parameter will contain, but maybe there's a difference?

**Comment by johnstt** [comment]
`PM_TruePointContents` gives the same result, yeah.
`PM_TraceTexture` always gives back a null string (passing the entity as 0 to get the world based on some other code I saw floating around..)

It's a good thing this game isn't open sourced, because I love spending like 3 hours on something that should take 5 minutes.

**Comment by SamVanheer** [comment]
You could replicate `PM_TruePointContents`, fortunately it basically amounts to this:
```
int PM_TruePointContents(vec3_t p)
{
  if ( !pmove->physents[0].model )
    return CONTENTS_EMPTY;

    return PM_HullPointContents(pmove->physents[0].model->hulls, pmove->physents[0].model->hulls[0].firstclipnode, p);
}
```

Where `PM_HullPointContents` is this: https://github.com/id-Software/Quake/blob/master/QW/client/pmovetst.c#L93

It appears to be identical, so at least this will let you check the contents value directly in a debugger.

**Comment by johnstt** [comment]
Tried that, it lets me see that the contents value really is -1...

So my maps don't have these parts labelled as sky.
It seems like checking the texture should be guaranteed to work - after all, it does render using the sky texture in game... maybe I need to reimplement some part of that to see why it's giving me a null string.

**Comment by SamVanheer** [comment]
How are you using `PM_TraceTexture`? It's not the most intuitive, you have to specify 2 points between which a traceline will be made, the first surface hit is the one whose texture is returned.
This code shows how to use it: https://github.com/ValveSoftware/halflife/blob/master/dlls/player.cpp#L3625

The engine function will use the same underlying code on the server, and equivalent code on the client. You're probably passing the trace end position, which probably doesn't touch the actual surface.

**Comment by johnstt** [comment]
Yeah, I'm passing it the same 2 vectors I use for the original trace (the one that finds a wall to jump off - it works next to a skybox but not in midair so it's _finding something_..)

**Comment by SamVanheer** [comment]
Try using endpos as the starting position, and endpos + VectorNormalize( endpos - old startpos ) * 16 as the end position. It should get you the sky texture at that point.

**Comment by johnstt** [comment]
That seems to work! ..but only rarely. I'll have to experiment a bit more, but something is funny about how often this code (under `PM_Jump`) runs and how rarely all the walljump code runs like I expect...

Thanks for the help so far! Now to unwind some of my other movement code and get to the bottom of this..
